### PR TITLE
SHiP move client handling off the main thread

### LINKED ISSUE #627: refactor thread and asio usage in state history plugin
Some usage of asio in state history plugin looks at odds with typical asio patterns. For example: needing a connection to call an explicit `close()` function after a failure occurs (normally you'd just let the shared_ptr fall out of scope to dtor it), and some sort of `stopping` variable (normally you'd just `stop()` the io_context which then ensures no callbacks are running, and then just let everything dtor itself). To be fair there may be yet-undiscovered-to-me reasons these are required, but so far I haven't seen it.

state history plugin is also very loose and haphazard with how it handles its thread, oftentimes punting back and forth between main thread. Yet, it seems like state history plugin's thread usage could be awfully simplistic: build up blocks' traces & deltas on the main thread and then when `on_accepted_block()` simply `post()` over to ship's thread which then _does everything it needs to entirely on the ship thread without ever needing to go back to main thread_.

Resolving the above issues will make it significantly easier to reason about state history plugin's flow and will be less prone to introducing failures.

### PR DESCRIPTION
**Pull request by heifner**
Moves client processing of `state_history_plugin` requests off the main application thread. 
- `block_log` has been made thread safe.
- `net_plugin` updated to use thread safe `block_log` from net threads instead of having to post onto the main thread for processing.
- SHiP writes to the state history logs on the main thread, but reads from them on the SHiP thread. The state history logs have been made thread safe.

Verified this branch does not crash under heavy load #592 
Resolves #627

### CODE REVIEW

File: libraries/chain/block_log.cpp [hunk 1]
```diff
@@ -1184,46 +1186,79 @@ namespace eosio { namespace chain {
...
    void     block_log::set_initial_version(uint32_t ver) { detail::block_log_impl::default_initial_version = ver; }
-   uint32_t block_log::version() const { return my->version(); }
+   uint32_t block_log::version() const {
+      std::lock_guard g(my->mtx);
+      return my->version();
+   }
 
-   void block_log::append(const signed_block_ptr& b, const block_id_type& id) { my->append(b, id, fc::raw::pack(*b)); }
+   void block_log::append(const signed_block_ptr& b, const block_id_type& id) {
+      std::vector<char> packed_block = fc::raw::pack(*b);
```
Comment (linh2931) [inline libraries/chain/block_log.cpp:1200]: Should this line be protected too? Would someone modify `b` while `pack` is running?

File: libraries/chain/controller.cpp [hunk 2]
```diff
@@ -392,14 +392,15 @@ struct controller_impl {
...
       EOS_ASSERT( fork_db.root(), fork_database_exception, "fork database not properly initialized" );
 
-      const auto& log_head = blog.head();
+      block_id_type log_head_id = blog.head_id();
+      const bool valid_log_head = !log_head_id.empty();
 
-      auto lib_num = log_head ? log_head->block_num() : (blog.first_block_num() - 1);
+      auto lib_num = valid_log_head ? block_header::num_from_id(log_head_id) : (blog.first_block_num() - 1);
 
       auto root_id = fork_db.root()->id;
```
Comment (linh2931) [inline libraries/chain/controller.cpp:400]: You can add `const` to other local variables in this function as they are not modified after initialized.

File: libraries/chain/controller.cpp [hunk 3]
```diff
@@ -3120,11 +3121,6 @@ block_state_ptr controller::fetch_block_state_by_number( uint32_t block_num )con
 } FC_CAPTURE_AND_RETHROW( (block_num) ) }
```
Comment (linh2931) [inline libraries/chain/controller.cpp:3122]: Maybe add a comment saying this is thread-safe.

File: libraries/chain/block_log.cpp (hunk 1 above)
Comment (heifner) [reply libraries/chain/block_log.cpp:1200]: We would have to protect all access to signed_block to guard against that. We already assume that signed_block is not modified after finalization. We already use signed_block on other threads after finalization.

File: libraries/chain/controller.cpp (hunk 3 above)
Comment (heifner) [reply libraries/chain/controller.cpp:3122]: There is a comment in the header that says that.

File: libraries/chain/controller.cpp (hunk 2 above)
Comment (heifner) [reply libraries/chain/controller.cpp:400]: Done.

**Review by linh2931 — APPROVED** [review]

**Review by 766C6164 — APPROVED** [review]

**Comment by greg7mdp** [comment]
@heifner I've not started reviewing this yet, and I see you have 2 approvals, so feel free to merge it if you'd like. I don't mind having a look either if you'd like.

**Comment by heifner** [comment]
> @heifner I've not started reviewing this yet, and I see you have 2 approvals, so feel free to merge it if you'd like. I don't mind having a look either if you'd like.

This has to wait on PR #628 which is waiting on PR #532. You can review if you like, I'm good either way.

**Pull request merged**